### PR TITLE
Charge handling -- Partial charges settings and charge transformation support

### PR DESCRIPTION
**Pull request by ijpulidos**
This set of changes aim to support charge transformations (up to one unit of charge changes) and give the users the ability to specify the partial charge assigning method via a friendly settings object. This is based on a lot of the functionality already included in the `RelativeHybridTopologyProtocol` in the `openfe` package, with some minor modifications.

Solves #19 
Solves #47 
Closes #20

### CODE REVIEW

**Comment by ijpulidos** [comment]
This adds a `charge.py` module to the `utils` subpackage. This might be open to suggestion on whether having such module and utilities is the better way of doing this. I just found a little confusing the way these things are organized on the `openfe` repo, since there are things in different places. I just found convenient to have them in a single module/place as possible, if that makes any sense.

**Comment by ijpulidos** [comment]
Actually, I noticed that we are not really exhaustively testing neither the partial charge assignation settings and variants, nor the handling of alchemical charge transformations. I think we need to add a minimal set of tests for these changes. Please suggest tests that we could add here, as possible.

**Comment by ijpulidos** [comment]
From the meeting it was agreed that we should port the host-guest system sanity check to this code base and set up a self-hosted runner on AWS (or similar) to run them, when possible.

**Comment by ijpulidos** [comment]
I just added the host-guest charge-changing transformation validation test. I created a `test_sanity.py` module for this, since I didn't know where to put it or even how to name it. I accept suggestions for this.

@hannahbaumann As discussed in our previous meeting, can you please take a look at the [host-guest test](https://github.com/choderalab/feflow/blob/80a843ec0decd64df1a47391e40466fcdde9b625/feflow/tests/test_sanity.py#L60) and see if that seems reasonable? Thanks!. For now it's using the default settings for the protocol and it's running in ~4 hours in an A40 gpu (we could probably lower this time if we optimize the number of steps to get the convergence we need), the default runs 100 cycles of 12500 steps of 4ps.

I also marked the test as `slow` so it would only get run when `--runslow` is passed as an argument flag to `pytest` because we will be looking into running these tests on our self-hosted runners through GHA, once that's set up.

**Review by hannahbaumann — COMMENTED** [review]
The host-guest test looks good to me, only thing I was wondering whether there should also be a check that the error is not too high, since at a high error (forward_reverse_sum_error) the test will likely pass but the results are maybe not meaningful.

File: feflow/protocols/nonequilibrium_cycling.py [hunk 1]
```diff
@@ -198,26 +225,26 @@ def _execute(self, ctx, *, state_a, state_b, mapping, settings, **inputs):
...
-                # due to issues with partial charge generation in ambertools
-                # we default to using the input conformer for charge generation
-                off_mol.assign_partial_charges(
-                    "am1bcc", use_conformers=off_mol.conformers
-                )
+        # Assign partial charges to all small mols
+        all_openff_mols = list(
+            chain(all_alchemical_mols.values(), common_small_mols.values())
+        )
+        self._assign_openff_partial_charges(
```
Comment (IAlibay) [inline feflow/protocols/nonequilibrium_cycling.py:232]: I generally suggest adding a test like this to Protocol to ensure that user changes (which are the recommended way in which charges are going to be passed to this Protocol) work: https://github.com/OpenFreeEnergy/openfe/blob/main/openfe/tests/protocols/test_openmm_equil_rfe_protocols.py#L708

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just two things but otherwise lgtm!

**Comment by ijpulidos** [comment]
As per the previous dev meeting, we decided that we also want to test the different options for assigning partial charges are working as expected in the code. This is a WIP, I'm writing the tests for this.

File: feflow/protocols/nonequilibrium_cycling.py (hunk 1 above)
Comment (ijpulidos) [reply feflow/protocols/nonequilibrium_cycling.py:233]: This is a good idea, I'll be working on that asap. I added some tests to check that the actual values for the methods and backends are being passed and used, but a test on the actual assigned charges should work better.

Comment (ijpulidos) [reply feflow/protocols/nonequilibrium_cycling.py:233]: From meeting we want this to be more in terms of ensuring user charges are not changed anywhere in the protocol.

File: pyproject.toml [hunk 2]
```diff
@@ -63,8 +63,9 @@ where = ["."]
 # Ref https://setuptools.pypa.io/en/latest/userguide/datafiles.html#package-data
 [tool.setuptools.package-data]
 feflow = [
-    "py.typed"
+    "py.typed",
 ]
+"feflow.data.host-guest" = ["*.json"]
```
[the remark this answers is not in the document]
Comment (ijpulidos) [reply pyproject.toml]: I solved it in a different way that is a bit more automatic, every version controlled file in packages will be added. I think this is a sane default.

File: feflow/protocols/nonequilibrium_cycling.py (hunk 1 above)
Comment (ijpulidos) [reply feflow/protocols/nonequilibrium_cycling.py:235]: Should be getting tested now in https://github.com/choderalab/feflow/blob/aec9428d1a943d156cedf3fe299f23545575058a/feflow/tests/test_nonequilibrium_cycling.py#L462

**Comment by ijpulidos** [comment]
@IAlibay This should be ready for new review now. Thanks!

**Review by IAlibay — APPROVED** [review]
Looks good to me! Thanks @ijpulidos !

**Pull request merged**